graph/memstore/quadstore.go

```diff
@@ -117,17 +116,17 @@ func (qs *QuadStore) ApplyDeltas(deltas []graph.Delta, ignoreOpts graph.IgnoreOp
 		case graph.Add:
 			if !ignoreOpts.IgnoreDup {
 				if _, exists := qs.indexOf(d.Quad); exists {
-					return graph.ErrQuadExists
+					return &graph.QuadError{Action: d.Action, Quad: d.Quad, Err: graph.ErrQuadExists}
```

Comment:
There are three bad actions. Doing it that way would either require having three distinct error types or alternatively the error string of QuadError could be dependent on the Action field: +1 => "quad exists", -1 => "quad does not exist", other => "invalid".

The approach you are suggesting was what I originally was considering (actually the even simpler approach of `type QuadError quad.Quad`). The approach here mimics os.PathError.